### PR TITLE
Delete artifacts not in the template 

### PR DESCRIPTION
**Pull request by lordozb**

### CODE REVIEW

File: build_and_deploy/clients/artifacts_client.ts [hunk 1]
```diff
@@ -93,14 +93,24 @@ export class ArtifactClient {
...
+        const baseUrl: string = this.getBaseurl(workspace, environment, resourceType);
+        let param: Params = await getParams(true, environment);
+        let token = param.bearer;
+        return await this.artifactDeletionTask(baseUrl, resourceType, payload, token);
+    }
 
-    public async WaitForAllDeployments() {
-        for (let i = 0; i < this.deploymentTrackingRequests.length; i++) {
+    public async WaitForAllDeployments(isDelete: boolean){
+        for(let i=0;i<this.deploymentTrackingRequests.length;i++){
```
Comment (praveenmathamsetty) [inline build_and_deploy/clients/artifacts_client.ts:104]: Nit: indentation

File: build_and_deploy/clients/artifacts_client.ts [hunk 2]
```diff
@@ -244,21 +254,18 @@ export class ArtifactClient {
...
                 let resStatus = res.message.statusCode;
-                SystemLogger.info(`ArtifactDeploymentTask status: ${resStatus}; status message: ${res.message.statusMessage}`);
+                SystemLogger.info(`For Artifact: ${payloadObj.name}: ArtifactDeploymentTask status: ${resStatus}; status message: ${res.message.statusMessage}`);
 
                 if (resStatus != 200 && resStatus != 201 && resStatus != 202) {
                     // Remove this after testing
                     res.readBody().then((body) => {
                         if (!!body) {
                             let responseJson = JSON.parse(body);
-                            SystemLogger.info("Deploy artifact failed: " + JSON.stringify(responseJson));
```
Comment (praveenmathamsetty) [inline build_and_deploy/clients/artifacts_client.ts:261]: why we removed this line, at-least print status code?

File: build_and_deploy/orchestrator.ts [hunk 3]
```diff
@@ -40,7 +47,22 @@ export class Orchestrator {
...
+            {
+                // Delete extra artifacts in the workspace
+                SystemLogger.info("Attempting to delete artifacts from workspace, that were not in the template.");
+                var artifactsInWorkspace = await getArtifactsFromWorkspace(this.targetWorkspace, this.environment);
+                SystemLogger.info(`Found ${artifactsInWorkspace.length} artifacts in the workspace.`);
+                var artifactsToDeleteInWorkspace = getArtifactsToDeleteFromWorkspace(artifactsInWorkspace, artifactsToDeploy, typeMap);
+                SystemLogger.info(`Found ${artifactsToDeleteInWorkspace.length} artifacts in the workspace that many need to be deleted.`);
+                var artifactsToDeleteInWorkspaceInOrder = getArtifactsToDeleteFromWorkspaceInOrder(artifactsToDeleteInWorkspace);
+                await this.deleteResourcesInOrder(this.artifactClient, artifactsToDeleteInWorkspaceInOrder!, this.targetWorkspace, this.environment, armParameterContent);
+                SystemLogger.info("Completed deleting artifacts from workspace, that were not in the template.");
```
Comment (praveenmathamsetty) [inline build_and_deploy/orchestrator.ts:60]: any exception handling needed here? incase of deleting artifacts failed due to what ever reason

File: build_and_deploy/orchestrator.ts [hunk 4]
```diff
@@ -52,7 +74,17 @@ export class Orchestrator {
...
             await this.deployBatch(artifactClient, batchOfArtifacts, targetWorkspace, environment);
-            await artifactClient.WaitForAllDeployments();
+            await artifactClient.WaitForAllDeployments(false);
+        }
+    }
+
+
+    private async deleteResourcesInOrder(artifactClient: ArtifactClient, artifactsToDelete: Resource[][],
+          targetWorkspace: string, environment: string, armParameterContent: string) {
+        for(let i=0;i<artifactsToDelete.length;i++){
```
Comment (praveenmathamsetty) [inline build_and_deploy/orchestrator.ts:84]: indentation

File: build_and_deploy/utils/workspace_artifacts_getter.ts [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+import * as httpInterfaces from 'typed-rest-client/Interfaces';
+import { checkIfArtifactExists, checkIfNameExists, Resource } from './arm_template_utils';
+import { Artifact } from './artifacts_enum';
+import {SystemLogger} from "./logger";
+
+const userAgent: string = 'synapse-github-cicd-deploy-task'
+const requestOptions: httpInterfaces.IRequestOptions = {};
+const client: httpClient.HttpClient = new httpClient.HttpClient(userAgent, undefined, requestOptions);
+
+const artifactTypesToQuery:Artifact[] = [
```
Comment (praveenmathamsetty) [inline build_and_deploy/utils/workspace_artifacts_getter.ts:13]: Is this list not exists in another file? pls cross check try reuse if already exists

File: build_and_deploy/orchestrator.ts (hunk 3 above)
Comment (lordozb) [reply build_and_deploy/orchestrator.ts:60]: This is already inside a try block

File: build_and_deploy/utils/workspace_artifacts_getter.ts (hunk 5 above)
Comment (lordozb) [reply build_and_deploy/utils/workspace_artifacts_getter.ts:13]: We have an enum which is used here. We don't have a list

File: build_and_deploy/clients/artifacts_client.ts (hunk 2 above)
Comment (lordozb) [reply build_and_deploy/clients/artifacts_client.ts:261]: Not sure why was this removed. Yes, it makes sense to keep it there to print the failure

**Review by praveenmathamsetty — APPROVED** [review]

**Pull request merged**